src/app/signalhandler.cpp

```diff
@@ -0,0 +1,145 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "signalhandler.h"
+
+#include <QtGlobal>
+
+#include <csignal>
+#include <tuple>
+
+#ifndef Q_OS_WIN
+#ifndef Q_OS_HAIKU
+#include <unistd.h>
+#endif
+#elif defined DISABLE_GUI
+#include <io.h>
+#endif
+
+#include <QCoreApplication>
+
+#include "base/version.h"
+
+#ifdef STACKTRACE
+#ifdef Q_OS_UNIX
+#include "stacktrace.h"
+#else
+#include "stacktrace_win.h"
+#ifndef DISABLE_GUI
+#include "stacktracedialog.h"
+#endif // DISABLE_GUI
+#endif // Q_OS_UNIX
+#endif //STACKTRACE
+
+namespace
+{
+    // sys_signame[] is only defined in BSD
+    const char *const sysSigName[] =
+    {
+#if defined(Q_OS_WIN)
+        "", "", "SIGINT", "", "SIGILL", "", "SIGABRT_COMPAT", "", "SIGFPE", "",
+        "", "SIGSEGV", "", "", "", "SIGTERM", "", "", "", "",
+        "", "SIGBREAK", "SIGABRT", "", "", "", "", "", "", "",
+        "", ""
+#else
+        "", "SIGHUP", "SIGINT", "SIGQUIT", "SIGILL", "SIGTRAP", "SIGABRT", "SIGBUS", "SIGFPE", "SIGKILL",
+        "SIGUSR1", "SIGSEGV", "SIGUSR2", "SIGPIPE", "SIGALRM", "SIGTERM", "SIGSTKFLT", "SIGCHLD", "SIGCONT", "SIGSTOP",
+        "SIGTSTP", "SIGTTIN", "SIGTTOU", "SIGURG", "SIGXCPU", "SIGXFSZ", "SIGVTALRM", "SIGPROF", "SIGWINCH", "SIGIO",
+        "SIGPWR", "SIGUNUSED"
+#endif
+    };
+
+#if !(defined Q_OS_WIN && !defined DISABLE_GUI) && !defined Q_OS_HAIKU
+    void reportToUser(const std::initializer_list<const char *> args)
+    {
+        const auto signalSafePrint = [](const char *str)
+        {
+            const size_t strLen = strlen(str);
+#ifdef Q_OS_WIN
+            if (_write(_fileno(stderr), str, strLen) < static_cast<int>(strLen))
+                std::ignore = _write(_fileno(stdout), str, strLen);
+#else
+            if (write(STDERR_FILENO, str, strLen) < static_cast<ssize_t>(strLen))
+                std::ignore = write(STDOUT_FILENO, str, strLen);
+#endif
+        };
+
+        for (const char *arg : args)
+            signalSafePrint(arg);
+    }
+#endif
+
+    void normalExitHandler(const int signum)
+    {
+#if !(defined Q_OS_WIN && !defined DISABLE_GUI) && !defined Q_OS_HAIKU
+        const char msg1[] = "Catching signal: ";
+        const char msg2[] = "\nExiting cleanly\n";
+        reportToUser({msg1, sysSigName[signum], msg2});
```

Comment:
I would just declare `safePrint()` function and call it here (and there) directly (as many times as needed). It doesn't look like `reportToUser()` adds anything useful, but it looks redundant to me.

Reply:
The idea was to provide variadic arguments, or at least in a similar way.

Reply:
IMO, it doesn't make sense if it doesn't add any argument handling, but just calls some function for each of them. It turns out that you're just packing arguments so that you can just unpack them right away.

Reply:
Done. 

>but just calls some function for each of them

Utilized std::for_each instead of writing repetitive calls (which I find it silly).

Reply:
> Utilized std::for_each instead of writing repetitive calls (which I find it silly).

IMO, you complicate trivial things. But at least it doesn't look so confusing now.

Reply:
>IMO, you complicate trivial things. 

Not at all, it resembles to the commonly used `map` operation better.